text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,182 @@
+- Feature Name: guaranteed_slice_repr
+- Start Date: 2025-02-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC guarantees the in-memory representation of slice and str references.
+Specifically, `&[T]` and `&mut [T]` are guaranteed to have the same layout as:
+
+```rust
+#[repr(C)]
+struct Slice<T> {
+    data: *const T,
+    len: usize,
+}
+```
+
+The layout of `&str` is the same as that of `&[u8]`, and the layout of
+`&mut str` is the same as that of `&mut [u8]`.
+
+# Motivation
+[motivation]: #motivation
+
+This RFC allows non-Rust (e.g. C or C++) code to read from or write to existing
+slices and to declare slice fields or locals.
+
+For example, guaranteeing the representation of slices allows non-Rust code to
+read from the `data` or `len` fields of `string` in the type below without
+intermediate FFI calls into Rust:
+
+```rust
+#[repr(C)]
+struct HasString {
+    string: &'static str,
+}
+```
+
+Note: prior to this RFC, the type above is not even properly `repr(C)` since the
+size and alignment of slices were not guaranteed. However, the Rust compiler
+accepts `repr(C)` declaration above without warning.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Slices are represented with a pointer and length pair. Their in-memory layout is
+the same as a `#[repr(C)]` struct like the following:
+
+```rust
+#[repr(C)]
+struct Slice<T> {
+    data: *const T,
+    len: usize,
+}
+```
+
+The precise ABI of slices is not guaranteed, so `&[T]` may not be passed by-value
```

Comment:
Wny not go ahead and define this ABI as well?

Reply:
Our ABI for this type is not the ABI of the equivalent C struct.

Reply:
In particular, its return ABI falls outside what can be obtained via C code on some platforms.

Reply:
Understood, thanks for explaining. That's.. really unfortunate.

One *could* still rigorously define the ABI to be what rustc does right now, and let callers in other languages use per-architecture inline assembly if they want to (this should be possible right?). But that's a much harder sell, and it's reasonable to leave this out of this RFC